system/lib/wasmfs/js_file_backend.cpp

```diff
@@ -0,0 +1,106 @@
+
+#include "backend.h"
+#include "wasmfs.h"
+
+namespace wasmfs {
+
+using js_index_t = uint32_t;
+
+// This class describes a file that lives in JS Memory
+class JSFile : public DataFile {
+  // This index indicates the location of the JS File in $wasmFS$JSMemoryFiles.
+  js_index_t index;
+
+  // JSFiles will write from a Wasm Memory buffer into a JS array defined in
+  // $wasmFS$JSMemoryFiles.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    EM_ASM(
+      {
+        // Resize the typed array if the length of the write buffer exceeds its
+        // capacity.
+        var size = wasmFS$JSMemoryFiles[$0].length;
+        if ($2 >= size) {
+          var oldContents = wasmFS$JSMemoryFiles[$0];
+          wasmFS$JSMemoryFiles[$0] = new Uint8Array($2);
+          wasmFS$JSMemoryFiles[$0].set(oldContents.subarray(0, $2));
+        }
+        wasmFS$JSMemoryFiles[$0].set(HEAPU8.subarray($1, $1 + $2), $3);
+      },
+      index,
+      buf,
+      len,
+      (size_t)offset);
+
+    return __WASI_ERRNO_SUCCESS;
+  }
+
+  // JSFiles will read from JS Memory into a Wasm Memory buffer.
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    assert(offset + len - 1 < getSize());
+    EM_ASM({ HEAPU8.set(wasmFS$JSMemoryFiles[$0].subarray($1, $2), $3); },
+           index,
+           (size_t)offset,
+           (size_t)offset + len,
+           buf);
+
+    return __WASI_ERRNO_SUCCESS;
+  }
+
+  // The size of the JSFile is defined as the length of the JS array in
+  // $wasmFS$JSMemoryFiles.
+  size_t getSize() override {
+    return (size_t)EM_ASM_INT({ return wasmFS$JSMemoryFiles[$0].length; },
+                              index);
+  }
+
+public:
+  JSFile(mode_t mode) : DataFile(mode) {
+    // Add the new JSFile to the $wasmFS$JSMemoryFiles array and assign the
+    // array index.
+    index = (js_index_t)EM_ASM_INT({
+      wasmFS$JSMemoryFiles.push(new Uint8Array(0));
+      return wasmFS$JSMemoryFiles.length - 1;
+    });
+  }
+
+  // Remove the typed array file contents in $wasmFS$JSMemoryFiles array.
+  ~JSFile() {
+    EM_ASM({ wasmFS$JSMemoryFiles[$0] = null; }, index);
+  }
+
+  class Handle : public DataFile::Handle {
+
+    std::shared_ptr<JSFile> getFile() { return file->cast<JSFile>(); }
+
+  public:
+    Handle(std::shared_ptr<File> dataFile) : DataFile::Handle(dataFile) {}
+  };
+
+  Handle locked() { return Handle(shared_from_this()); }
+};
+
+class JSFileBackend : public Backend {
+
+public:
+  std::shared_ptr<DataFile> createFile(mode_t mode) override {
+    return std::make_shared<JSFile>(mode);
+  }
+  std::shared_ptr<Directory> createDirectory(mode_t mode) override {
+    return std::make_shared<Directory>(mode, this);
+  }
+};
+
+// This function is exposed to users to instantiate a new JSBackend.
+extern "C" backend_t create_js_file_backend() {
+  wasmFS.backendTable.push_back(std::make_unique<JSFileBackend>());
+  return wasmFS.backendTable.back().get();
```

Comment:
Does this need locking so that the table doesn't change in between the `push_back` and the `back()`? It might be better to do the `get` first to avoid this issue (but the table will still need locking to ensure accesses are synchronized).

Reply:
I think this is fixed in the previous PR, with the `addBackend()` that returns a `backend_t` instead. It has a lock on the vector.